vg/VG_Line.3
.\" Copyright (c) 2008-2022 Julien Nadeau Carriere <vedge@csoft.net>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 21, 2022
.Dt VG_LINE 3
.Os Agar 1.7
.Sh NAME
.Nm VG_Line
.Nd agar-vg line entity
.Sh SYNOPSIS
.Bd -literal
#include <agar/core.h>
#include <agar/gui.h>
#include <agar/vg.h>
.Ed
.Sh DESCRIPTION
.\" IMAGE(/widgets/VG_Line.png, "A VG_Line")
.Nm
is a
.Xr VG 3
node class which describes a line segment in terms of two points.
.Sh INTERFACE
.nr nS 1
.Ft "VG_Line *"
.Fn VG_LineNew "VG_Node *parent" "VG_Point *p1" "VG_Point *p2"
.Pp
.Ft "void"
.Fn VG_LineThickness "VG_Line *vl" "Uint8 thickness"
.Pp
.Ft "void"
.Fn VG_LineStipple "VG_Line *vl" "Uint16 pattern"
.Pp
.Ft "void"
.Fn VG_LineEndpointStyle "VG_Line *vl" "enum vg_endpoint_style style" "..."
.Pp
.nr nS 0
The
.Fn VG_LineNew
function creates a new line segment entity attached to
.Fa parent .
The
.Fa p1
and
.Fa p2
arguments define the endpoints.
.Pp
.Fn VG_LineThickness
defines the thickness of the line in pixels (default = 1 pixel).
Note that the line thickness parameter does not influence proximity queries
used by "select" tools. it only affects rendering, so very thick lines should
be rendered as
.Xr VG_Polygon 3
instead.
.Pp
.Fn VG_LineStipple
defines a 16-bit, OpenGL-style stipple pattern (default is 0xffff).
The order in which bits are used is undefined and depends on order of
rendering.
.Pp
By default, thick lines are rendered using a square endpoint style.
.Fn VG_LineEndpointStyle
specifies the style of endpoints to use in rendering the line.
Acceptable values include:
.Bd -literal
.\" SYNTAX(c)
enum vg_line_endpoint {
	VG_LINE_SQUARE,		/* Square endpoint */
	VG_LINE_BEVELED,	/* Beveled endpoint */
	VG_LINE_ROUNDED,	/* Rounded endpoint (circular) */
	VG_LINE_MITERED		/* Mitered endpoint */
};
.Ed
.Pp
If
.Dv VG_LINE_MITERED
is used,
.Fn VG_LineEndpointStyle
expects a following
.Ft Uint8
argument describing the miter length in pixels.
.Sh SEE ALSO
.Xr VG 3 ,
.Xr VG_Point 3
.Sh HISTORY
The
.Nm
class first appeared in Agar 1.3.3.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - YYYY-MM-DD
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**AG_Driver**](https://libagar.org/man3/AG_Driver): New operation: `setMouseAutoCapture` (enable, disable or reset mouse auto-capture on button-down).
- [**AG_DriverSw**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Error**](https://libagar.org/man3/AG_Error): New function `AG_Debug2()` (output debug message when debug level is >= 2).
- [**AG_FontBf**](https://libagar.org/man3/AG_FontBf): New bitmap font engine for the `.agbf` font format.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required. Improved appearance of `AG_MenuView`.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): When an `AG_MenuView` is detached, immediately invalidate the `view` pointers of its `AG_MenuItem`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Text**](https://libagar.org/man3/AG_Text): The surface returned by `AG_TextRender()` now includes Ascent information under `Guide 0`.
- [**AG_Text**](https://libagar.org/man3/AG_Text): New function `AG_TextRenderRTL()` for rendering text right-to-left.
- [**AG_Text**](https://libagar.org/man3/AG_Text): New functions `AG_FontGetStyleName()` and `AG_FontGetStyleByName()`. Convert between the `Uint` and string representations of font styles (including styles, weights and width variants).
- [**AG_Text**](https://libagar.org/man3/AG_Text): New function `AG_FontGetFamilyStyles()`. Return an array of `AG_Font` style flags describing every style, weight and width variant available in a given font's family.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) and [**AG_Editable**](https://libagar.org/man3/AG_Editable): Implement Undo/Redo functionality.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Draw lines to express the relationship between tree items.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistScrollToSelection()`.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_EXPAND_NODES`. Display nodes (items with children) initially in expanded state.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_NO_SCALE_ICON`. Disable icon scaling when the icon width exceeds the standard width.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_FIXED_HEIGHT`. Don't reset the item height on "font-changed".
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New item flag `AG_TLIST_ITEM_DISABLED`. Disable the item individually (prevent selection and display in "disabled" state).
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Improve keyboard navigation. Allow RIGHT/LEFT keys to Expand/Collapse node items.
- [**AG_Event**](https://libagar.org/man3/AG_Event): New macro [AG_EVENT_DUMP](https://libagar.org/man3/AG_EVENT_DUMP) to produce a listing of `argv[]` arguments on the console.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New flag `AG_WIDGET_DISABLE_ON_ATTACH` and functions `AG_PushDisabledState()` and `AG_PopDisabledState`. Make attached child widgets start in DISABLED mode (without raising "widget-disabled").
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New event _"padding-changed"_. Generated by the style compiler to signal that the padding attribute of a widget has changed.
- [**MAP**](https://libagar.org/man3/MAP): Introduce dynamic Map Objects with [**MAP_Object**](https://libagar.org/man3/MAP_Object). Map Objects coexist with static Map elements and are embedded with the serialized `MAP`. Nodes refer to `MAP_Object` instances using an intermediate `MAP_Location` structure (which adds per-node flags, z-coordinate, height, layer and neighbor information). More than one `MAP` node may refer to a given `MAP_Object`.
- [**MAP**](https://libagar.org/man3/MAP): New Insert Object (`InsertObj`) tool.
- [**MAP**](https://libagar.org/man3/MAP): New functions `MAP_DuplicateLocation()`, `MAP_NodeDelLocation()`, `MAP_NodeDelLocationAtIndex()`.
- [**MAP**](https://libagar.org/man3/MAP): Added height (`h`) and z-coordinate (`z`) attributes to `MAP_Item`. Introduce `MAP_Item` classes.
- [**MAP**](https://libagar.org/man3/MAP): Added Undo/Redo and History Buffer to Map Editor. New function `MAP_ClearHistory()`.
- [**MAP**](https://libagar.org/man3/MAP): Added validity tests to `MAP_Node` and `MAP_Item`.
- [**MAP**](https://libagar.org/man3/MAP): Added persistent Library to the Map Editor. The `pLibs` pointer of `MAP` may be used to specify an alternate `AG_Object` as VFS Root for the Library. Similarly, `pMaps` may be used to specify an alternate VFS Root for loaded Maps.
- `install-agartest.exe` installer for agartest on Windows.
- Provide copies of the OFL as separate files (OFL11.txt and LICENSE.ofl).
- Install a copy of the generated Makefile.config as ${DATADIR}/agar.mk.

vg/vg_line.c
/*
 * Copyright (c) 2004-2019 Julien Nadeau Carriere <vedge@csoft.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Line entity.
 */

#include <agar/core/core.h>
#include <agar/gui/widget.h>
#include <agar/gui/primitive.h>
#include <agar/gui/numerical.h>
#include <agar/gui/radio.h>
#include <agar/gui/iconmgr.h>
#include <agar/vg/vg.h>
#include <agar/vg/vg_view.h>
#include <agar/vg/icons.h>

#include <stdarg.h>

VG_Line *
VG_LineNew(void *pNode, VG_Point *p1, VG_Point *p2)
{
	VG_Line *vl = Malloc(sizeof(VG_Line));

	VG_NodeInit(vl, &vgLineOps);
	vl->p1 = p1;
	vl->p2 = p2;
	VG_AddRef(vl, p1);
	VG_AddRef(vl, p2);
	VG_NodeAttach(pNode, vl);
	return (vl);
}

void
VG_LineThickness(VG_Line *vl, Uint8 t)
{
	VG *vg = VGNODE(vl)->vg;

	AG_ObjectLock(vg);
	vl->thickness = t;
	AG_ObjectUnlock(vg);
}

void
VG_LineStipple(VG_Line *vl, Uint16 s)
{
	VG *vg = VGNODE(vl)->vg;

	AG_ObjectLock(vg);
	vl->stipple = s;
	AG_ObjectUnlock(vg);
}

static void
Init(void *_Nonnull obj)
{
	VG_Line *vl = obj;

	vl->p1 = NULL;
	vl->p2 = NULL;
	vl->endPt = VG_LINE_SQUARE;
	vl->stipple = 0xffff;
	vl->miterLen = 0;
	vl->thickness = 1;
}

void
VG_LineEndpointStyle(VG_Line *vl, enum vg_line_endpoint style, ...)
{
	VG *vg = VGNODE(vl)->vg;
	va_list ap;

	AG_ObjectLock(vg);
	
	vl->endPt = style;

	if (style == VG_LINE_MITERED) {
		va_start(ap, style);
		vl->miterLen = (Uint8)va_arg(ap, int);
		va_end(ap);
	}

	AG_ObjectUnlock(vg);
}

static int
Load(void *_Nonnull obj, AG_DataSource *_Nonnull ds, const AG_Version *_Nonnull ver)
{
	VG_Line *vl = obj;

	if ((vl->p1 = VG_ReadRef(ds, vl, "Point")) == NULL) {
		return (-1);
	}
	if ((vl->p2 = VG_ReadRef(ds, vl, "Point")) == NULL) {
		return (-1);
	}
	vl->endPt = (enum vg_line_endpoint)AG_ReadUint8(ds);
	vl->stipple = AG_ReadUint16(ds);
	vl->miterLen = AG_ReadUint8(ds);
	vl->thickness = AG_ReadUint8(ds);
	return (0);
}

static void
Save(void *_Nonnull obj, AG_DataSource *_Nonnull ds)
{
	VG_Line *vl = obj;

	VG_WriteRef(ds, vl->p1);
	VG_WriteRef(ds, vl->p2);
	AG_WriteUint8(ds, (Uint8)vl->endPt);
	AG_WriteUint16(ds, vl->stipple);
	AG_WriteUint8(ds, vl->miterLen);
	AG_WriteUint8(ds, vl->thickness);
}

static void
Draw(void *_Nonnull obj, VG_View *_Nonnull vv)
{
	VG_Line *vl = obj;
	AG_Color c = VG_MapColorRGB(VGNODE(vl)->color);
	int x1,y1, x2,y2;

	VG_GetViewCoords(vv, VG_Pos(vl->p1), &x1, &y1);
	VG_GetViewCoords(vv, VG_Pos(vl->p2), &x2, &y2);

	/* XXX TODO: endpoint style */
	AG_DrawLine(vv, x1,y1, x2,y2, &c);
}

static void
Extent(void *_Nonnull obj, VG_View *_Nonnull vv, VG_Vector *_Nonnull a,
    VG_Vector *_Nonnull b)
{
	VG_Line *vl = obj;
	VG_Vector p1, p2;

	p1 = VG_Pos(vl->p1);
	p2 = VG_Pos(vl->p2);
	a->x = MIN(p1.x, p2.x);
	a->y = MIN(p1.y, p2.y);
	b->x = MAX(p1.x, p2.x);
	b->y = MAX(p1.y, p2.y);
}

static float
PointProximity(void *_Nonnull obj, VG_View *_Nonnull vv, VG_Vector *_Nonnull vPt)
{
	VG_Line *vl = obj;
	VG_Vector v1 = VG_Pos(vl->p1);
	VG_Vector v2 = VG_Pos(vl->p2);

	return VG_PointLineDistance(v1, v2, vPt);
}

static void
Delete(void *_Nonnull obj)
{
	VG_Line *vl = obj;

	if (VG_DelRef(vl, vl->p1) == 0) { VG_Delete(vl->p1); }
	if (VG_DelRef(vl, vl->p2) == 0) { VG_Delete(vl->p2); }
}

static void
Move(void *_Nonnull obj, VG_Vector vCurs, VG_Vector vRel)
{
	/* TODO */
}

static void *_Nonnull
Edit(void *_Nonnull obj, VG_View *vv)
{
	VG_Line *vl = obj;
	AG_Box *box = AG_BoxNewVert(NULL, AG_BOX_EXPAND);
	const char *endPtStyles[] = {
		N_("Square"),
		N_("Beveled"),
		N_("Rounded"),
		N_("Mitered"),
		NULL
	};
	AG_Numerical *num;

	num = AG_NumericalNew(box, 0, NULL, _("Thickness: "));
	AG_BindUint8(num, "value", &vl->thickness);

	num = AG_NumericalNew(box, 0, NULL, _("Stipple pattern: "));
	AG_BindUint16(num, "value", &vl->stipple);

	AG_LabelNew(box, 0, _("Endpoint style: "));
	AG_RadioNewUint(box, AG_RADIO_EXPAND, endPtStyles, (void *)&vl->endPt);

	return (box);
}

VG_NodeOps vgLineOps = {
	N_("Line"),
	&vgIconLine,
	sizeof(VG_Line),
	Init,
	NULL,			/* destroy */
	Load,
	Save,
	Draw,
	Extent,
	PointProximity,
	NULL,			/* lineProximity */
	Delete,
	Move,
	Edit
};
